next.config.js

```diff
@@ -4,7 +4,31 @@ const path = require("path")
 const Dotenv = require("dotenv-webpack")
 
 const currentExport = {
-  webpack: config => {
+  pageExtensions: ["jsx", "js", "ts", "tsx"],
+  webpack: (config, options) => {
+    const { dir, defaultLoaders, dev, isServer } = options
+    config.resolve.extensions = [".ts", ".tsx", ".js", ".jsx", ".json"]
+
+    // HMR for TS pages
+    if (dev && !isServer) {
+      config.module.rules.push({
+        test: /\.(ts|tsx)$/,
+        loader: "hot-self-accept-loader",
+        include: [path.join(dir, "pages")],
+        options: {
+          extensions: /\.(ts|tsx)$/,
+        },
+      })
+    }
```

Comment:
this isn't working amazingly, maybe there's a better way, or maybe it is because I'm on a canary release

```js
const hmr = config.module.rules.filter(({ use = {} }) => use.loader === 'hot-self-accept-loader')
hmr.test = ...
```

Is an alternative